### PR TITLE
SCIONprobes

### PR DESCRIPTION
**Pull request by FR4NK-W**
This add support for SCION: [https://www.scion-architecture.net](https://www.scion-architecture.net/pages/faq/)

Adds SCION Probes sping and pping

Add Probes and Orders for SCION SCMP and pingpong tool

### CODE REVIEW

File: lib/Pingmachine/Order/PPing.pm [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+sub probe_instance_key {
+    my ($self) =@_;
+
+    my @keys;
+    push (@keys, "interval:".$self->interval)   if ($self->interval);
+    push (@keys, "source_ip:".$self->source_ip) if ($self->source_ip);
+    push (@keys, "interface:".$self->interface) if ($self->interface);
+    push (@keys, "v6:".$self->ipv6) if ($self->ipv6);
+    scalar @keys or @keys = ('');
```
Comment (kbucheli) [inline lib/Pingmachine/Order/PPing.pm]: Above line has no effect and can be removed:
```
$ perl -wE 'my @a; say defined(join("|", @a))'
1
```

File: lib/Pingmachine/Order/SPing.pm [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+sub probe_instance_key {
+    my ($self) =@_;
+
+    my @keys;
+    push (@keys, "interval:".$self->interval)   if ($self->interval);
+    push (@keys, "source_ip:".$self->source_ip) if ($self->source_ip);
+    push (@keys, "interface:".$self->interface) if ($self->interface);
+    push (@keys, "v6:".$self->ipv6) if ($self->ipv6);
+    scalar @keys or @keys = ('');
```
Comment (kbucheli) [inline lib/Pingmachine/Order/SPing.pm]: ditto

File: lib/Pingmachine/Probe/PPing.pm [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    # Determine $interval (fping -p)
+    my $interval; # interval applies only to periods between pings in series (fping -p)
+    if($self->interval) {
+        $interval = $self->interval;
+    }
+    else {
+        $interval = int(($step * 1000 * 0.8 - $hostcount*$MIN_WAIT - $TIMEOUT) / $pings);
+        $interval >= 1000 or
+            die "fping: calculated interval too small: $interval (step = $step, pings = $pings)\n";
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: Please adapt the error text as this is not an fping probe.

File: lib/Pingmachine/Probe/PPing.pm [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+
+    # Make sure that we can process all hosts
+    if($interval / $hostcount < $MIN_WAIT) {
+        die "fping: step * 1000 / (pings * hostcount) must be at least 10 (step=$step, pings=$pings, hostcount=$hostcount\n";
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: ditto

File: lib/Pingmachine/Probe/PPing.pm [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        '-interval', $interval.'ms',
+        '-count', $pings,
+        '-timeout', $TIMEOUT.'s',
+        '-format', '"%[3]s : %[5]s"',
+        '-remote', $single_host,
+        '-local', $self->source_ip,
+        '-aggregate',
+    ];
+    my $switch_dir = 'cd '.$scionDir;
+    my $cmd_string = $switch_dir." && ".join(" ", @$cmd);
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: Better
```
my $cmd_string = "cd $scionDir && "`. join(' ', @$cmd);
```

File: lib/Pingmachine/Probe/PPing.pm [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+sub _kill_current_job {
+    my ($self) = @_;
+
+    # Kill pping, if still running
+    my $job = $self->current_job;
+    if($job->{pid}) {
+        if(kill(0, $job->{pid})) {
+            $log->warning("killing unfinished pping process (step: ".$self->step.", pings: ".$self->pings.", offset: ".$self->time_offset().")");
+            kill 9, $job->{pid};
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: First it should be checked if it is really a pping process. Or that this is a child process of this process. Do not kill innocents...

File: lib/Pingmachine/Probe/PPing.pm [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    $log->debug("adding results") if $log->is_debug();
+
+    # Add results (to RRD)
+    if(scalar keys %results) {
+        my $now = int(AnyEvent->now);
+        my $step = $self->step;
+        my $rrd_time = $now - $step - $now%$step;
+        for my $host (keys %results) {
+            my $h2o = $job->{host2order}{$host};
+            if(not defined $h2o) {
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: Perl Best Practices: do not use `not`.
And the defined can also be omitted as a reference is expected and all forms of Perl false is false.

File: lib/Pingmachine/Probe/SPing.pm [hunk 8]
```diff
@@ -0,0 +1,244 @@
+package Pingmachine::Probe::SPing;
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/SPing.pm:1]: The same issues as for Pingmachine::Probe::PPing

File: lib/Pingmachine/Probe/PPing.pm [hunk 9]
```diff
@@ -0,0 +1,247 @@
+package Pingmachine::Probe::PPing;
+
+use Any::Moose;
+use AnyEvent;
+use AnyEvent::Util;
+use Log::Any qw($log);
+use List::Util qw(shuffle);
+
+my $scionDir = $ENV{'SC'}'
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: What if `$ENV{SC}` is not set?

File: lib/Pingmachine/Probe/PPing.pm [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+my $PPING_BIN = $scionDir.'/bin/pingpong_legacy';
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: Better make use of string interpolation
`"$scionDir/bin/pingpong_legacy"`

**Review by kbucheli — COMMENTED** [review]
Except for the kill there are only minor cosmetic issues.
Thanks!

File: lib/Pingmachine/Order/PPing.pm (hunk 1 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Order/PPing.pm]: Done

File: lib/Pingmachine/Order/SPing.pm (hunk 2 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Order/SPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 3 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 4 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 5 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 6 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Changed it to check that we kill the process of the command we started.

File: lib/Pingmachine/Probe/PPing.pm (hunk 7 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done
Didn't know about the unless keyword.

File: lib/Pingmachine/Probe/SPing.pm (hunk 8 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/SPing.pm:1]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 9 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Added a default value, environment variable $SC is set when following the standard SCION installation procedure.

File: lib/Pingmachine/Probe/PPing.pm (hunk 10 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

**Comment by FR4NK-W** [comment]
@ kbucheli, thank you for your code review
Should we add the same kill safeguards also the FPing Probe?

**Comment by kbucheli** [comment]
Yes please.

File: lib/Pingmachine/Probe/PPing.pm [hunk 11]
```diff
@@ -0,0 +1,257 @@
+package Pingmachine::Probe::PPing;
+
+use Any::Moose;
+use AnyEvent;
+use AnyEvent::Util;
+use Log::Any qw($log);
+use List::Util qw(shuffle);
+
+my $SCION_DIR = $ENV{'SC'} // '/opt/OSAGscion';
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: The default should be as in the upstream project and not as in a internal package name.

File: lib/Pingmachine/Probe/PPing.pm [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+sub _kill_current_job {
+    my ($self) = @_;
+
+    # Kill pping, if still running
+    my $job = $self->current_job;
+    my $job_pid = $job->{pid};
+    # Check that we are killing the process we started and not an innocent bystander
+    if($job_pid && (`/bin/ps -p $job_pid -o cmd=` == "".join(' ', @$effective_cmd))) {
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: Calling external tools is not very efficient and should be avoided if possible.
You can read "/proc/$job_pid/cmdline"

File: lib/Pingmachine/Probe/PPing.pm (hunk 11 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 12 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done
Please let me know if that is the right way to do it.

File: lib/Pingmachine/Probe/PPing.pm [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    my ($self) = @_;
+
+    # Kill pping, if still running
+    my $job = $self->current_job;
+    my $job_pid = $job->{pid};
+    if($job_pid) {
+        # Check that we are killing the process we started and not an innocent bystander
+        my $cmd_match = 0;
+        if (open(proc_fh, "/proc/${job_pid}/cmdline")) {
+            $cmd_match = (join("", <proc_fh>) == "".join(' ', @$effective_cmd));
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: This does a numeric comparison.
Better:
```
$cmd_match = (join('', readline(proc_fh)) eq join(' ', @$effective_cmd));
```
But what is $effective_cmd? A global variable! That should not be.

File: lib/Pingmachine/Probe/PPing.pm [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+        my $host = $order->pping->host;
+        push @{$job{host2order}{$host}}, $order;
+    }
+    $job{hostlist} = join("\n", shuffle keys %{$job{host2order}}) . "\n",
+    $self->current_job(\%job);
+
+    # Run pingpong
+    my $single_host = substr $job{hostlist}, 0, -1;
+    $log->info("single_host: ".$single_host);
+    my $cmd = [
```
Comment (kbucheli) [inline lib/Pingmachine/Probe/PPing.pm]: better do 
```
my @cmd = (
```
as it is used as array everywhere.

Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Done

File: lib/Pingmachine/Probe/PPing.pm (hunk 13 above)
Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: Changed the comparison

my $effective_cmd is file global and it gets set in _start_new_job and is reused by _kill_current_job.
Should I make $effective_cmd package global with the `our` keyword?
File global should be OK, no? `l. 11  +my $effective_cmd`

Comment (FR4NK-W) [reply lib/Pingmachine/Probe/PPing.pm]: I'll pass along the effective_cmd with the job, like pid is. $effective_cmd being global is indeed wrong.

**Review by kbucheli — APPROVED** [review]

**Pull request merged**